prime-router/src/main/kotlin/azure/ReportFunction.kt

```diff
@@ -103,152 +114,159 @@ class ReportFunction : Logging {
+        sender: Sender,
         context: ExecutionContext,
-        workflowEngine: WorkflowEngine
-    ) {
-        if (report != null && report.schema.topic.lowercase() == "covid-19") {
-            // next check that we're dealing with an external file
-            val clientSource = report.sources.firstOrNull { it is ClientSource }
-            if (clientSource != null) {
-                context.logger.info("Writing deidentified report data to the DB")
-                // wrap the insert into an exception handler
-                try {
-                    workflowEngine.db.transact { txn ->
-                        // verify the file exists in report_file before continuing
-                        if (workflowEngine.db.checkReportExists(report.id, txn)) {
-                            val deidentifiedData = report.getDeidentifiedResultMetaData()
-                            workflowEngine.db.saveTestData(deidentifiedData, txn)
-                            context.logger.info("Wrote ${deidentifiedData.count()} rows to test data table")
-                        } else {
-                            // warn if it does not exist
-                            context.logger.warning(
-                                "Skipping write to metadata table because " +
-                                    "reportId ${report.id} does not exist in report_file."
-                            )
-                        }
-                    }
-                } catch (pse: PSQLException) {
-                    // report this but move on
-                    context.logger.severe(
-                        "Exception writing COVID test metadata " +
-                            "for ${report.id}: ${pse.localizedMessage}.\n" +
-                            pse.stackTraceToString()
+        workflowEngine: WorkflowEngine,
+        actionHistory: ActionHistory
+    ): HttpResponseMessage {
+        // determine if we should be following the sync or async workflow
+        val isAsync = processingType(request) == ProcessingType.Async
+        val errors: MutableList<ResultDetail> = mutableListOf()
+        val warnings: MutableList<ResultDetail> = mutableListOf()
+        // The following is identical to waters (for arch reasons)
+        val validatedRequest = validateRequest(workflowEngine, request)
+        warnings += validatedRequest.warnings
+        handleValidation(validatedRequest, request, actionHistory, workflowEngine)?.let {
+            return it
+        }
+
+        var report = workflowEngine.createReport(
+            sender,
+            validatedRequest.content,
+            validatedRequest.defaults,
+            errors,
+            warnings
+        )
+
+        // if we are processing a message asynchronously, the report's next action will be 'process'
+        if (isAsync && report != null) {
+            report.nextAction = TaskAction.process
+        }
+
+        var response: HttpResponseMessage
```

Comment:
Good candidate for a tech-debt ticket. There's a more fluent builder pattern for responses that would probably help normalize this. Either way we should choose a common response building pattern for the system

Reply:
#3041